File: Depth-First Search (DFS)/110. Balanced Binary Tree.cpp

class Solution {
public:
    int hight(TreeNode*root)
    {
        if(!root) return 0;
        int l = hight(root->left);
        int r = hight(root->right);
        return max(l,r)+1;
    }
    bool isBalanced(TreeNode* root) 
    {
        if(!root) return true;
        if(!root->left and !root->right) return true;

        int l = hight(root->left);
        int r = hight(root->right);

        if(abs(l-r) > 1) return false;
        return isBalanced(root->left) and isBalanced(root->right);
    }
};

File: 110. Balanced Binary Tree.py
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        store = []
        def dfs(tree):
            if not tree:
                return 0
            l = dfs(tree.left)
            r = dfs(tree.right)
            store.append((l, r))
            return max(l, r) + 1
        dfs(root)

        for l, r in store:
            if abs(l - r) > 1:
                return False
        return True